l5kit/l5kit/sampling/agent_sampling.py

```diff
@@ -15,6 +15,94 @@
+    history_slice = get_history_slice(state_index, history_num_frames, history_step_size, include_current_state=True)
+    future_slice = get_future_slice(state_index, future_num_frames, future_step_size)
+
+    history_frames = frames[history_slice].copy()  # copy() required if the object is a np.ndarray
+    future_frames = frames[future_slice].copy()
+
+    sorted_frames = np.concatenate((history_frames[::-1], future_frames))  # from past to future
+
+    # get agents (past and future)
+    agent_slice = get_agents_slice_from_frames(sorted_frames[0], sorted_frames[-1])
+    agents = agents[agent_slice].copy()
+    # sync interval with the agents array
+    history_frames["agent_index_interval"] -= agent_slice.start
+    future_frames["agent_index_interval"] -= agent_slice.start
+    history_agents = filter_agents_by_frames(history_frames, agents)
+    future_agents = filter_agents_by_frames(future_frames, agents)
+
+    # get traffic lights (past and future)
+    tl_slice = get_tl_faces_slice_from_frames(sorted_frames[0], sorted_frames[-1])
+    tl_faces = tl_faces[tl_slice].copy()
+    # sync interval with the traffic light faces array
+    history_frames["traffic_light_faces_index_interval"] -= tl_slice.start
+    history_tl_faces = filter_tl_faces_by_frames(history_frames, tl_faces)
+    future_tl_faces = filter_tl_faces_by_frames(future_frames, tl_faces)
+
+    return history_frames, future_frames, history_agents, future_agents, history_tl_faces, future_tl_faces
+
+
+def compute_agent_velocity(
+    history_positions_m: np.ndarray, future_positions_m: np.ndarray, history_step_time: float, future_step_time: float
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Compute estimated velocities by finite differentiation on future positions as(pos(T+t) - pos(T))/t.
+    This simple approach gives less than 0.5% velocity difference
+    compared to (pos(T+t) - pos(T-t))/2t on v1.1/sample.zarr.tar.
+
+    Args:
+        history_positions_m (np.ndarray): history XY positions in meters
+        future_positions_m (np.ndarray): future XY positions in meters
+        history_step_time (np.ndarray): length of a step in second
+        future_step_time (np.ndarray): length of a step in second
+
+    Returns:
+        Tuple[np.ndarray, np.ndarray]: history and future XY speeds
+
+    """
+    # [future_num_frames, 2]
+    future_positions_diff_m = np.concatenate((future_positions_m[:1], np.diff(future_positions_m, axis=0)))
+    # [future_num_frames, 2]
+    future_vels_mps = np.float32(future_positions_diff_m / future_step_time)
+
+    # current position is included in history positions
+    # [history_num_frames, 2]
+    history_positions_diff_m = np.diff(history_positions_m, axis=0)
+    # [history_num_frames, 2]
+    history_vels_mps = np.float32(history_positions_diff_m / history_step_time)
+
+    return history_vels_mps, future_vels_mps
+
+
```

Comment:
Seems like the history_step_time here may be misused, right? Example, the user needs to know that history_step_size and history_step_time must be consistent with each other, otherwise the returned frames may represent different things. Or am I missing something?
Also, this is probably not something in scope for this PR either, but the way history_step_time and history_step_size are used seems like they are related with the dataset's frame rate.

Reply:
yes this is a very good point. `history_step_size` and `future_step_time` are not properly check for values greater than 1 in the codebase. I'm actually tempted to remove them as they can be misused

Reply:
#213 will remove that para